Interactive sim setup section (#1266)
ihmeuw/vivarium_research: docs/source/model_design/designing_vivarium_model/interactive_sim/index.rst

Setting up an Interactive Sim
-----------------------------

Setting up an interactive sim for the first time can be confusing. Follow these 
instructions along with the example notebook included for help. 
Another useful resource is the `Vivarium InteractiveContext documentation <https://vivarium.readthedocs.io/en/latest/api_reference/interface/interactive.html?highlight=InteractiveContext#vivarium.interface.interactive.InteractiveContext>`_. 

**Example notebook**

> This `notebook <https://github.com/ihmeuw/vivarium_research_nih_us_cvd/blob/main/interactive_sim_example_setup.ipynb>`_ includes all of the steps seen below using the CVD project as an example. 

Reviewer: I can see the notebook just fine, but it does have a bunch of really long warnings that make the notebook a little harder to navigate and understand for me - I haven't used an interactive sim before though so not sure if that's something you can change?

Author: Good point! Yes, I will add those lines and include how to do this in the steps as well.

> **Step 1: General Prep**

Reviewer 2: Should we make these subheadings? That would make them linkable, show up in the TOC, etc

>   - Ensure that you have cloned the engineer's GitHub repo and have pulled the latest version of the simulation.  

Reviewer: ```suggestion
  - Ensure that you have cloned the engineers' GitHub repo and have pulled the latest version of the simulation.  
```

  - Check that you have 'vivarium' installed in your environment and can import 'InteractiveContext' in a Jupyter notebook. 
  - If you need more general help with Git or environments for this part, see these :ref:`computing resources <computing>`.

Example code: 

:: 

  from vivarium import InteractiveContext 

**Step 2: Creating a Simulation** 

Now that you're prepped and able to run a simulation, we can actually load the sim! To 
do this, we will use the 'InteractiveContext' function in Vivarium. 

First, you'll need to locate the model specs file. It will be in the engineering repo usually 
in a file path similar to: :code:`<project_repo_name>/src/<project_repo_name>/model_specifications/`. 
It will be a .yaml file. 

If you can't locate the proper model specs file or are unsure - ask the engineers for help! 

> Once you have that, you can follow this code to create a simulation 

Reviewer: ```suggestion
Once you have that, you can follow this code to create a simulation: 
```

::

  path = Path('<file_path_for_model_spec_file>')
  

>   sim = InteractiveContext(Path(path), setup=False)

Reviewer 3: ```suggestion
  sim = InteractiveContext(path, setup=False)
```
I don't think you need to make it into a path here, it is already a path from the previous line. (But please confirm that this works!)

> This creates an object called 'sim' that is the simulation. 

Reviewer 3: ```suggestion
This creates an object called `sim` that is the simulation. 
```
cooler typesetting

**Step 3: Updating the Simulation Parameters**

Often, you will want to use slightly different parameters than are the standard in the model 
spec. For example, you might want to change the population size. To do this, you can use 
'configuration.update'. 

Below is an example of changing the population size from 100,000 in the model spec file 
to be 50,000 for the interactive sim. You can use this to help change anything in the 
'configuration' section of the model spec. 

These are the lines we will be editing as written in the model spec file: 

:: 

  configuration: 
    population: 
      population_size: 100_000
      age_start: 5 
      age_end: 125 

Now, we will update the 'population_size' variable to be 50,000 in our notebook: 

::

>   sim.configuration.update({
>                           'population':
>                               {'population_size': 50_000,
>                               },})

Reviewer 2: ```suggestion
  sim.configuration.update({
    'population':  {
      'population_size': 50_000,
    },
  })
```

Just formatting things nicer (please test!)

Instead of using 'configuration.update' you can just directly change the values in the 
model spec file. However, since this file exists in the engineering repo, changing it on 

> your local machine can cause errors and confusion. Also, whenever you pull from main, 
> your updates would be removed. Therefore, we do NOT recommend this approach. 

Reviewer 2: ```suggestion
your local machine can cause errors and confusion.
Therefore, we do NOT recommend this approach for simple changes like
updating the population size.
However, editing the model spec file will be necessary for more complex changes
such as removing components and observers from the simulation.
```

Pulling wouldn't overwrite changes unless there were merge conflicts.

And, I think there are some things that can't be done from `sim.configuration.update` --
am I right about this @aflaxman?

I like to have this as a cell in my interactive sim notebooks, so that any changes I've made to
files are preserved in the notebook output and I know what I had changed if I look at the notebook
later:

```
! git -C <path to engineering repo> show -s
! git -C <path to engineering repo> diff HEAD
```

Author: Hey - yeah good callout. My understanding is that you can only use this approach to change things in the "configuration" section of the model spec, which isn't everything. I have tried to make that clear here. I basically punted the larger scale editing to the next section and am going to leave this as just the basics for now. But this is helpful info with the logging of changes - I didn't know that! I'll be sure to include it below.

**Step 4: Loading Data from the Simulation** 

> Now that you have the parameters set-up, you're ready to start getting data from the simulation. 

Reviewer: ```suggestion
Now that you have the parameters set up, you're ready to start getting data from the simulation. 
```

The first step is to run 'sim.setup'. Running this command will take some time and possibly generate 
some pink warning text. Don't worry! Just wait for the cell to finish running. 

Once it's done, you will have a simulated population. You can use 'get_population' to create a dataset 
with your population. Some simulant data automatically gets recorded for your sim. To find a list of these, 
list the columns in your dataset. Other simulant data does not automatically get added, but can also 
be saved. To find a list of additional sim data available, use 'list_values'. If you find 
something from the list that you want included in your data, just add it using 'get_value'. 

The below will show using all of these in practice: 

:: 

  sim.setup() # Sets up the simulation 
  pop0 = sim.get_population() # Generates a dataset with some simulant data included 
  pop0.columns # Lists the columns in your simulant dataset 
  sim.list_values() # Lists the additional columns you can add to the dataset 

>   sim.get_value('<variable_from_list_values>')(pop0.index) #Pulls data for all simulants 

Reviewer 2: Should this show how to join it into the pop0 dataframe? That's what I would typically do.
